### ISSUE TITLE
Add support for haproxy PROXY protocol

### ISSUE THREAD

**Issue by farcaller** [znc/znc#915]
It would be nice to have haproxy [PROXY protocol](http://www.haproxy.org/download/1.5/doc/proxy-protocol.txt) support, so that znc gets real client IP addresses if haproxy is used to forward client connections to znc instance.

**Comment by DarthGandalf** [comment]
Do you mean, as an alternative to X-Forwarded-For?

**Comment by farcaller** [comment]
I mean, for IRC client connections.

**Comment by kylef** [comment]
I feel this is outside of the scope of ZNC and is better suited as a module for ZNC.

**Comment by farcaller** [comment]
I thought ZNC modules have no API to intercept the prologue from incoming TCP socket, no? Haproxy basically prepends client connection details when it connects to ZNC, so this must be handled even earlier than TLS.

**Comment by stevesbrain** [comment]
I presume @DarthGandalf that X-Forwarded-For is currently implemented then, and works for the web interface - but does it work for IRC connections as well?

I imagine if not, then PROXY protocol as mentioned above would be useful for providing the real IP in the actual TCP connection (similar to what happens with Dovecot).

**Comment by kylef** [comment]
I believe support for `X-Forwarded-For` header was implemented in #349.

**Comment by DarthGandalf** [comment]
@kylef This is about analog of `X-Forwarded-For` for any TCP connection, not just HTTP.

**Comment by elyscape** [comment]
This would most likely need to be supported by the upstream CSocket library.

**Comment by elyscape** [comment]
>This would most likely need to be supported by the upstream CSocket library.

I have opened jimloco/Csocket#74 to track this.

**Comment by Mikaela** [comment]
This could be nice for people exposing their ZNC instance as an onion service, because Tor supports exporting the circuit ID as IPv6 address using haproxy's proxy protocol, which may be *a bit* more useful than `127.0.0.1`.

       HiddenServiceExportCircuitID protocol
           The onion service will use the given protocol to expose the global circuit identifier of each inbound client circuit via the selected protocol.
           The only protocol supported right now 'haproxy'. This option is only for v3 services. (Default: none)

           The haproxy option works in the following way: when the feature is enabled, the Tor process will write a header line when a client is
           connecting to the onion service. The header will look like this:

           "PROXY TCP6 fc00:dead:beef:4dad::ffff:ffff ::1 65535 42\r\n"

           We encode the "global circuit identifier" as the last 32-bits of the first IPv6 address. All other values in the header can safely be ignored.
           You can compute the global circuit identifier using the following formula given the IPv6 address "fc00:dead:beef:4dad::AABB:CCDD":

           global_circuit_id = (0xAA << 24) + (0xBB << 16) + (0xCC << 8) + 0xDD;

           In the case above, where the last 32-bit is 0xffffffff, the global circuit identifier would be 4294967295. You can use this value together with
           Tor’s control port where it is possible to terminate a circuit given the global circuit identifier. For more information about this see
           controls-spec.txt.

           The HAProxy version 1 proxy protocol is described in detail at https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt

**Comment by DarthGandalf** [comment]
This will be easier to implement after #1639 is finished

**Comment by glyph** [comment]
Any update on this? I've had to reorganize my personal infrastructure onto a cloud provider who doesn't allow for multiple IPs on a single machine so I'm feeling this pain every day :).

**Comment by Juliaria08** [comment]
If possible it'd also be a good idea to allow this to work with [certauth](https://wiki.znc.in/Certauth), which I guess it would anyways?
